tools/src/main/java/org/apache/kafka/trogdor/workload/ProduceBenchWorker.java

```diff
@@ -0,0 +1,337 @@
+            this.startMs = startMs;
+        }
+
+        @Override
+        public void onCompletion(RecordMetadata metadata, Exception exception) {
+            long now = Time.SYSTEM.milliseconds();
+            long durationMs = now - startMs;
+            sendRecords.recordDuration(durationMs);
+        }
+    }
+
+    public class SendRecords implements Callable<Void> {
+        private final Histogram histogram;
+
+        private final Throttle throttle;
+
+        private final Future<?> statusUpdaterFuture;
+
+        SendRecords() {
+            this.histogram = new Histogram(5000);
+            this.throttle = new Throttle(spec.targetMessagesPerSec());
+            this.statusUpdaterFuture = executor.scheduleWithFixedDelay(
+                new StatusUpdater(histogram), 1, 1, TimeUnit.MINUTES);
+        }
+
+        @Override
+        public Void call() throws Exception {
+            long startTimeMs = Time.SYSTEM.milliseconds();
+            try {
+                Properties props = new Properties();
+                props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, spec.bootstrapServers());
+//                props.put(ProducerConfig.BATCH_SIZE_CONFIG, 16384);
+//                props.put(ProducerConfig.ACKS_CONFIG, 1);
+//                props.put(ProducerConfig.LINGER_MS_CONFIG, 0);
+//                props.put(ProducerConfig.SEND_BUFFER_CONFIG, 128 * 1024);
+//                props.put(ProducerConfig.RECONNECT_BACKOFF_MS_CONFIG, 50);
+//                props.put(ProducerConfig.BUFFER_MEMORY_CONFIG, 32 * 1024 * 1024L);
+                for (Map.Entry<String, String> entry : spec.producerConf().entrySet()) {
+                    props.setProperty(entry.getKey(), entry.getValue());
+                }
+                byte[] key = new byte[512];
+                byte[] value = new byte[512];
+                KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(
+                            props, new ByteArraySerializer(), new ByteArraySerializer());
+                Future<RecordMetadata> future = null;
+                try {
+                    for (int m = 0; m < spec.maxMessages(); m++) {
+                        for (int i = 0; i < spec.activeTopics(); i++) {
+                            ProducerRecord<byte[], byte[]> record = new ProducerRecord<>(topicIndexToName(i), key, value);
+                            future = producer.send(record, new SendRecordsCallback(this, Time.SYSTEM.milliseconds()));
+                        }
+                        throttle.increment();
+                    }
+                } finally {
+                    if (future != null) {
+                        future.get();
+                    }
+                    producer.close();
+                }
+            } catch (Exception e) {
+                abort("SendRecords", e);
+            } finally {
+                statusUpdaterFuture.cancel(false);
+                long numSamples = 0;
+                synchronized (histogram) {
```

Comment:
Any reason `Histogram` itself doesn't have synchronized methods?

Reply:
Good point-- I made the methods in Histogram thread-safe.  That should also make this slightly more efficient since we can drop the lock when it's not needed.